### PR TITLE
feat(openid4vc): jwt format and more crypto

### PR DESCRIPTION
**Pull request by TimoGlastra**
Adds support for receiving JWT credentials in the openid4vc client, and also extends support for supported proof of possession JWA signature algorithms. 

To be able to support multiple signature algs, I had to remove the `kid` parameter, as the crypto that will be used, and the did methods the issuer supports has a big impact of the `kid` that will be selected.

For this, a new `proofOfPossessionVerificationMethodResolver` function MUST be supplied when requesting a credential through the openi4vc-client in AFJ, that is called whenever a verification method id (kid) must be determined for a to be created proof of possesion. The callback takes the following options, which should give the caller enough  information to determine the verification method:

```ts
export interface ProofOfPossessionVerificationMethodResolverOptions {
  /**
   * The credential format that will be requested from the issuer.
   * E.g. `jwt_vc` or `ldp_vc`.
   */
  credentialFormat: SupportedCredentialFormats

  /**
   * The JWA Signature Algorithm that will be used in the proof of possession.
   * This is based on the `allowedProofOfPossessionSignatureAlgorithms` passed
   * to the request credential method, and the supported signature algorithms.
   */
  proofOfPossessionSignatureAlgorithm: JwaSignatureAlgorithm

  /**
   * This is a list of verification methods types that are supported
   * for creating the proof of possession signature. The returned
   * verification method type must be of one of these types.
   */
  supportedVerificationMethods: string[]

  /**
   * The key type that will be used to create the proof of possession signature.
   * This is related to the verification method and the signature algorithm, and
   * is added for convenience.
   */
  keyType: KeyType

  /**
   * The credential type that will be requested from the issuer. This is
   * based on the credential types that are included the credential offer.
   */
  credentialType: string

  /**
   * Whether the issuer supports the `did` cryptographic binding method,
   * indicating they support all did methods. In most cases, they do not
   * support all did methods, and it means we have to make an assumption
   * about the did methods they support.
   *
   * If this value is `false`, the `supportedDidMethods` property will
   * contain a list of supported did methods.
   */
  supportsAllDidMethods: boolean

  /**
   * A list of supported did methods. This is only used if the `supportsAllDidMethods`
   * property is `false`. When this array is populated, the returned verification method
   * MUST be based on one of these did methods.
   *
   * The did methods are returned in the format `did:<method>`, e.g. `did:web`.
   */
  supportedDidMethods: string[]
}
```

It adds a bit of complexity for the caller, but there's simply no way for AFJ to know which did method or key we should use in a proof of possession, as that's heavily reliant on business logic. And passing it upfront also doesn't work, as the caller is missing important information. 

The crypto that can be used for the proof of possesion is based on what is supported by the wallet and the registered JSON-LD signatuer suites. So if we were to add new key types / JWA siganture algortihms or signature suites, the openid4vc client wouldn't have to be changed at all.

You can limit which algs and credential formats you want to allow, so those aren't considered as option when the issuer supports them through the `allowedCredentialFormats` and `allowedProofOfPossessionSignatureAlgorithms` properties. (e.g. I only want to receive `jwt_vc` credentials with `ES256` alg, even if my wallet and the issuer both support `Ed25519Signature2018`. We may want to also add a property for `allowedDidMethods` so we won't consider an option if your specific did method, or all did methods arern't supported by the issuer.

~~Dependant on #1468~~

### CODE REVIEW

**Comment by TimoGlastra** [comment]
For just the changes from this PR, see https://github.com/TimoGlastra/aries-framework-javascript/compare/feat/jwt-vc...TimoGlastra:aries-framework-javascript:feat/openid4vci-jwt-cyprot?expand=1

File: packages/openid4vc-client/src/OpenId4VcClientService.ts [hunk 1]
```diff
@@ -255,69 +167,320 @@ export class OpenId4VcClientService {
...
+        )
+      }
+
+      const jws = await this.jwsService.createJwsCompact(agentContext, {
+        key,
+        payload,
+        protectedHeaderOptions: {
+          alg: jwt.header.alg,
+          kid: jwt.header.kid,
+        },
```
Comment (blu3beri) [inline packages/openid4vc-client/src/OpenId4VcClientService.ts:480]: I ran into an issue with this. the JWT contained a `typ` header, which is required for openid4vci, but it was discarded from the header by doing it like this. Would it be possible to just pass along all the headers?

File: packages/openid4vc-client/src/OpenId4VcClientServiceOptions.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+/**
+ * The credential formats that are supported by the openid4vc client
+ */
+export type SupportedCredentialFormats = Extract<ClaimFormat, 'jwt_vc' | 'ldp_vc'>
+
+/**
+ * Options that are used for the pre-authorized code flow.
+ */
+export interface PreAuthCodeFlowOptions {
```
Comment (blu3beri) [inline packages/openid4vc-client/src/OpenId4VcClientServiceOptions.ts:11]: Loving all the docstrings!

**Review by blu3beri — APPROVED** [review]
Some very small comments. Already looking forward to resolving the merge conflicts regarding the openid4vc-client package :).

File: packages/core/src/crypto/jose/JoseHeader.ts
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/core/src/crypto/jose/JoseHeader.ts:1]: Nope, I deleted it in this PR (although github shows it as empty file)

File: packages/openid4vc-client/src/OpenId4VcClientService.ts [hunk 3]
```diff
@@ -255,69 +167,320 @@ export class OpenId4VcClientService {
...
+      const issuerSupportedCryptographicSuites = credentialFormat.cryptographic_suites_supported ?? []
+      // FIXME: somehow the MATTR Launchpad returns binding_methods_supported instead of cryptographic_binding_methods_supported
+      const issuerSupportedBindingMethods: string[] =
+        credentialFormat.cryptographic_binding_methods_supported ?? credentialFormat.binding_methods_supported ?? []
+
+      // For each of the supported algs, find the key types, then find the proof types
+      const signatureSuiteRegistry = agentContext.dependencyManager.resolve(SignatureSuiteRegistry)
+
+      let potentialSignatureAlgorithm: JwaSignatureAlgorithm | undefined
+      if (potentialCredentialFormat === 'jwt_vc') {
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/openid4vc-client/src/OpenId4VcClientService.ts:340]: It is already an if statement. Do you mean only if statements, and not if/else?

[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/openid4vc-client/src/OpenId4VcClientService.ts:340]: removed the else if in favour of a new if. LEt me know if that is what you meant

File: packages/openid4vc-client/src/OpenId4VcClientService.ts (hunk 1 above)
Comment (TimoGlastra) [reply packages/openid4vc-client/src/OpenId4VcClientService.ts:480]: Yeah good catch. I pass all values now, and throw an error if there's values we don't support for this flow (such as `x5c` and `jwk`)

**Comment by TimoGlastra** [comment]
> Some very small comments. Already looking forward to resolving the merge conflicts regarding the openid4vc-client package :).

Yeah that's going to be tedious :(

File: packages/openid4vc-client/src/OpenId4VcClientService.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (blu3beri) [reply packages/openid4vc-client/src/OpenId4VcClientService.ts]: I ment a switch statement, my bad. Was looking at the word if while typing.

**Pull request merged**